doc/tags_param.md

```diff
@@ -0,0 +1,248 @@
+## Tag based parameter framework
+
+### Table of Contents
+
+- [Motivation](#motivation)
+- [What are Tags?](#what-are-tags)
+- [API](#api)
+    - [sets() / gets()](#sets--gets)
+    - [has()](#has)
+    - [list_params()](#list_params)
+- [Python API](#python-api)
+- [For Developers](#for-developers)
+    - [BaseTag](#basetag)
+    - [Any](#any)
+    - [register_param() / register_member()](#register_param--register_member)
+    - [SWIG interface](#swig-interface)
+    - [Tests](#tests)
+
+### Motivation
+
+With the new framework, we aimed for the following:
+- A consistent API to set or get the value of a parameter in any shogun class, and allow the removal of innumerable setters and getters.
+- Cleanup the Shogun base classes and modularize things for easier development.
+- Work towards allowing plugins for Shogun.
+- Faster SWIG compilation with less symbols being exported.
+
+### What are Tags?
+
+`Tag` is a Shogun class which stores name and type information of a parameter. This allows a `Tag` object to be used as an identifier for a parameter in this new framework. Eg:
+
+```cpp
+auto width_tag = shogun::Tag<float64_t>("log_width");
+// width_tag can be used as an identifier for a parameter of
+// type float64_t and name "log_width"
+```
+
+### API
+
+#### sets() / gets()
+
+Let's start with an example.
+
+```cpp
+// We instantiate a Gaussian Kernel with width = 5
+float64_t width = 5;
+auto gkernel = shogun::GaussianKernel(width);
+// Update the width to 6
+gkernel.sets(width_tag, 6);
```

Comment:
`list_param()` is not available in the API because I faced some issues with SWIG while implementing it. Will update the readme when `list_param()` is merged in shogun.